apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/java/adapter/AdapterCommon.kt

```diff
@@ -103,9 +109,19 @@ internal fun readFromResponseCodeBlock(
   /**
    * Read the synthetic properties
    */
-  val checkTypename = if (syntheticProperties.isNotEmpty()) {
+  val typenameCodeBlock = if (syntheticProperties.any { it.requiresTypename }) {
     checkedProperties.add(__typename)
-    CodeBlock.of("$T.checkFieldNotMissing($__typename, $S);", JavaClassNames.Assertions, __typename)
+
+    CodeBlock.builder()
+        .apply {
+          if (regularProperties.none { it.info.responseName == "__typename" }) {
+            // We are in a nested fragment that needs access to __typename, get it from the buffered reader
+            add("$reader.rewind();\n")
+            add(javaTypenameFromReaderCodeBlock())
+          } else {
+            add("$T.checkFieldNotMissing($__typename, $S);", JavaClassNames.Assertions, __typename)
+          }
+        }.build()
```

Comment:
These changes are leftovers from https://github.com/apollographql/apollo-kotlin/pull/3930 that were not added to the Java codegen